worker/storageprovisioner/storageprovisioner.go

```diff
@@ -359,6 +375,28 @@ func (w *storageProvisioner) loop() error {
 	}
 }
 
+// processDependantChanges processes changes from a watcher strings channel. If
+// there are any changes, it calls the given function, repeating until there are
+// no more changes.
+// If there are no changes, it returns with no error.
+func (w *storageProvisioner) processDependantChanges(ctx *context, source watcher.StringsChannel, fn func(*context, []string) error) error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case changes, ok := <-source:
+			if !ok {
+				return errors.New("watcher closed")
+			}
+			if err := fn(ctx, changes); err != nil {
+				return errors.Trace(err)
+			}
+		default:
+			return nil
```

Comment:
maybe this should be after a timeout?

Reply:
What would be an adequate timeout in this scenario, as this would impact all deployments wanting to attach storage?

I could live with 1-second max, as that feels like a lifetime in this scenario.